### PR TITLE
vs_pass_fixes

### LINKED ISSUE #577: Explicit registration of passes required
Punch line: MS-VCR build does not comply with UD behavior of GCC for static initialization

On Windows 10 with VS2015Update2, git2+ and cmake3+ in PATH, I built binaryen:

```cmd
git clone https://github.com/WebAssembly/binaryen.git --recursive
cd binaryen
cmake .
cmake --build . --config Release
```

When I tried running the hello world example: `bin/asm2wasm ./test/hello_world.asm.js` in Windows PowerShell, it gives me: `Fatal: Could not find pass: dce`.

Apparently, `dce` and many other passes do not get initialized; only print, print-minimized and print-full get attention of static initializer. This is because the compiler/linker are optimizing those code paths out as suspected by @kripken: https://github.com/WebAssembly/binaryen/pull/576#issuecomment-224125210.

I have confirmed this by placing `#pragma comment(linker, "/INCLUDE:wasm::registerPass")` after [asm2wasm.cpp#L29](https://github.com/WebAssembly/binaryen/blob/e4a81f9/src/tools/asm2wasm.cpp#L29) and using `/OPT:NOREF` switch, the linker throws:

> LNK2001	unresolved external symbol wasm::registerPass	asm2wasm

Similar case study: http://stackoverflow.com/q/9459980 (see Edit section of original post)

Interesting comments:

> 
* Relying on static initialization to register functions is probably a bit error prone in the long run. Have you considered just doing explicit registration?
* I think if you want something initialized, try to initialize them MANUALLY other than depending on compiler's work

### PR DESCRIPTION
**Pull request by juj**
Add initialization functions for passes to avoid missing pass registration due to linker dead code elimination. Fixes #577.

### CODE REVIEW

File: src/passes/CoalesceLocals.cpp [hunk 1]
```diff
@@ -595,8 +595,14 @@ void CoalesceLocalsWithLearning::pickIndices(std::vector<Index>& indices) {
 
 // declare passes
 
-static RegisterPass<CoalesceLocals> registerPass1("coalesce-locals", "reduce # of locals by coalescing");
```
Comment (dschuff) [inline src/passes/CoalesceLocals.cpp]: If we are removing all the `RegisterPass` calls, can we now just delete its implementation? I don't see that in this PR.

**Comment by kripken** [comment]
I'm not too happy with this, was hoping to avoid the LLVM-like `create*` method which means each new pass needs boilerplate in at least 2 places (`passes.h` and `binaryen-shell.cpp`). But we don't seem to have a better solution here.

Except for making passes a shared library as opposed to static. Would that not fix the problem on windows?

File: src/passes/CoalesceLocals.cpp (hunk 1 above)
Comment (kripken) [inline src/passes/CoalesceLocals.cpp]: Yeah, if we go with this PR, then all the PassRegistry stuff can go away too.

File: src/passes.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/*
```
Comment (kripken) [inline src/passes.h]: might be nicer to put this file in `src/passes/`

**Comment by dschuff** [comment]
@kripken I just posted about the question of whether we should do this at all on #577. In particular I think we want to support both shared and static use cases.

File: src/passes/CoalesceLocals.cpp (hunk 1 above)
Comment (juj) [inline src/passes/CoalesceLocals.cpp]: Looks like PassRegistry is still needed, since the command line options want to map pass name strings to passes. Adjusted the registry to allow this.

**Comment by juj** [comment]
Updated pull request. Unrelated, had to add a `std::` to compile on VS2015.

**Comment by juj** [comment]
Travis complains about the build when running the tests, but I'm not sure if I understand what's going on. Trying to carefully read through to find if I typoed a pass name or something, but nothing catches my eye.

File: src/tools/binaryen-shell.cpp [hunk 3]
```diff
@@ -77,8 +77,8 @@ static void run_asserts(size_t* i, bool* checked, Module* wasm,
   std::unique_ptr<ShellExternalInterface> interface;
   std::unique_ptr<ModuleInstance> instance;
   if (wasm) {
-    interface = make_unique<ShellExternalInterface>();
-    instance = make_unique<ModuleInstance>(*wasm, interface.get());
+    interface = wasm::make_unique<ShellExternalInterface>();
```
Comment (kripken) [inline src/tools/binaryen-shell.cpp]: we do `using namespace wasm`, why are these needed? ambiguity somehow I guess? is there some other make_unique in the global namespace?

**Comment by kripken** [comment]
I think what's going on is that that example does not use any passes. So it was fine before - we don't register any, and don't call any. But now you added that header which declares them, and the method calling it exists, so the passes must exist. This means you need to link in `src/passes/*.cpp`.

Maybe we can just link the c++ example with libbinaryen.so like we do with the c examples. Right now in check.py we handle the two differently (look for `'example'`).

If this turns out to be messy, we can probably just remove the one c++ example we have. It's not very good, and we added a C api exactly so we'd have a stable api for people. I'd say if it's not easy to link to libbinaryen, let's get rid of it.

**Comment by kripken** [comment]
Merged #593 which contained this + some fixes.

One thing not resolved is what I asked about in this PR with the `wasm::` stuff.

File: src/tools/binaryen-shell.cpp (hunk 3 above)
Comment (juj) [inline src/tools/binaryen-shell.cpp]: VS2015 says there's `std::make_unique` that is ambiguous with this, and fails to compile. The project defines its own `wasm::make_unique` in `src/support/utilities.h` that it conflicts with.

Comment (dschuff) [inline src/tools/binaryen-shell.cpp]: I'm fine with making this `wasm::make_unique`. IIRC, `std::make_unique` is a new addition in C++14. Does VS2015 default to C++14?

Comment (jasonwilliams200OK) [inline src/tools/binaryen-shell.cpp]: > Does VS2015 default to C++14?

Yes, and `make_unique` is present since VS2013 (June 28, 2013 https://blogs.msdn.microsoft.com/vcblog/2013/06/28/c1114-stl-features-fixes-and-breaking-changes-in-vs-2013/)

I can send a PR with AppVeyor to verify the changes against VS2015 Update 3 if this is acceptable.

Comment (kripken) [inline src/tools/binaryen-shell.cpp]: That would be great, @jasonwilliams200OK :)

**Comment by kripken** [comment]
That other PR was merged so I think this one can be closed.
